In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random

# Домашнее задание к уроку 1

## Задача 1

Подберите скорость обучения (alpha) и количество итераций:

## Задача 2

В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее.
(дополнительная): вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию начинает изменяться ниже определенного порога (упрощенный аналог параметра tol в линейной регрессии в sklearn)

In [2]:
#Пространство признаков для объекта (стаж репетиторы)
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [3]:
# Целевая переменная (средний балл ЕГЭ)
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [4]:
n = X.shape[1]
W = np.array([1, 0.5])
dist = np.inf
eps = 1e-6
m = 10000
W

array([1. , 0.5])

In [5]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

## Два метода подбора значения $\alpha$

###  Метод случайного выбора 

In [37]:
best_i, best_a, best_err = m , 0, 100
for r in range(20):
    # Переношу определенние переменных в цикл, чтобы обойти ошибку переполнения
    X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])
    y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]  
    n = X.shape[1]
    W = np.array([1, 0.5])
    dist = np.inf
    eps = 1e-8
    m = 10000
    alpha = random.uniform(1e-1, 1e-7)
    for i in range(m):
        y_pred = np.dot(W, X)
        err = calc_mse(y, y_pred)
        W -= (alpha * (1/n * 2 * (X @ (y_pred - y))))
        if i % 100 == 0:
            print(i, W, err, alpha)
        if abs(dist - err) < eps:
            print(i, W, err, alpha)
            if i < best_i :
                best_i, best_a , best_err= i, alpha, err
            break
        if np.isinf(err):
            print("Ошибка переполнения")
            break
        dist = err
print(best_i, best_a, best_err)

0 [ 4.31296848 10.72149985] 3173.15 0.030063234840357445
100 [45.44814725  4.24014272] 47.760748152818316 0.030063234840357445
200 [47.15764311  3.92447123] 45.940679566108784 0.030063234840357445
300 [47.22903174  3.91128878] 45.93750554485172 0.030063234840357445
357 [47.23163382  3.91080828] 45.93750014844223 0.030063234840357445
0 [ 9.166505   25.69611342] 3173.15 0.07410621594854375
100 [3.88668857e+09 2.10480775e+10] 4.529135539726231e+21 0.07410621594854375
200 [5.73293396e+18 3.10462847e+19] 9.853935577293896e+39 0.07410621594854375
300 [8.45617845e+27 4.57938161e+28] 2.143898002384064e+58 0.07410621594854375
400 [1.24730120e+37 6.75466844e+37] 4.664429362840022e+76 0.07410621594854375
500 [1.83979121e+46 9.96325479e+46] 1.0148291223150609e+95 0.07410621594854375
600 [2.71372441e+55 1.46959761e+56] 2.2079402803340954e+113 0.07410621594854375
700 [4.00279125e+64 2.16768233e+65] 4.803764667691805e+131 0.07410621594854375
800 [5.90418751e+73 3.19736957e+74] 1.0451439827472429e+150

/home/egor/anaconda3/envs/py3-jupyter/lib/python3.8/site-packages/numpy/core/_methods.py:151: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(arr, axis, dtype, out, keepdims)
<ipython-input-5-bd4c0e5d829e>:2: RuntimeWarning: overflow encountered in square
  err = np.mean((y - y_pred)**2)


 [46.61137667  4.02534351] 46.162920214489716 0.039750464421365525
200 [47.22302773  3.91239746] 45.93754860288083 0.039750464421365525
272 [47.23170646  3.91079487] 45.93750011140476 0.039750464421365525
0 [ 6.13549262 16.34453259] 3173.15 0.04660156645349083
100 [46.93988259  3.96468238] 45.98821713138632 0.04660156645349083
200 [47.23010729  3.91109017] 45.93750246028637 0.04660156645349083
233 [47.23174763  3.91078727] 45.93750009274915 0.04660156645349083
0 [1.84898859 3.11938402] 3173.15 0.0077040706460443364
100 [27.66001917  7.52485665] 255.22111615052876 0.0077040706460443364
200 [38.47537746  5.527718  ] 87.83101071778984 0.0077040706460443364
300 [43.31427777  4.63417805] 54.32356610753144 0.0077040706460443364
400 [45.47925053  4.23439926] 47.61618730872503 0.0077040706460443364
500 [46.44788116  4.05553421] 46.273532538301076 0.0077040706460443364
600 [46.88125627  3.97550818] 46.004765574839475 0.0077040706460443364
700 [47.07515266  3.93970373] 45.950964938786456 0.00770

3000 [2.53993901e+099 1.37548540e+100] 2.534252513964433e+201 0.06893126478809841
3100 [5.08448657e+102 2.75346653e+103] 1.0155411323888333e+208 0.06893126478809841
3200 [1.01781986e+106 5.51192903e+106] 4.0695383979722333e+214 0.06893126478809841
3300 [2.03748649e+109 1.10338591e+110] 1.6307702607391087e+221 0.06893126478809841
3400 [4.07866986e+112 2.20877384e+113] 6.534922104768046e+227 0.06893126478809841
3500 [8.16474019e+115 4.42155535e+116] 2.6187138644551077e+234 0.06893126478809841
3600 [1.63442947e+119 8.85113331e+119] 1.0493870001734125e+241 0.06893126478809841
3700 [3.27182451e+122 1.77183264e+123] 4.2051676247648436e+247 0.06893126478809841
3800 [6.54958553e+125 3.54688015e+126] 1.6851204321616517e+254 0.06893126478809841
3900 [1.31110548e+129 7.10019588e+129] 6.752717428351028e+260 0.06893126478809841
4000 [2.62458988e+132 1.42132746e+133] 2.7059901355929277e+267 0.06893126478809841
4100 [5.25394191e+135 2.84523382e+136] 1.0843608801374591e+274 0.06893126478809841
4200 [1

Получаем после 20 подходов лучшее $i = 233$ и $\alpha  = 0.0466015$ и $err = 45.9375000$ 

### Второй способ, потихоньку уменьшить число 

In [43]:
best_i, best_a = m , 0
alpha = 1e-1
for r in range(20):
    # Переношу определенние переменных в цикл, чтобы обойти ошибку переполнения
    X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])
    y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]  
    n = X.shape[1]
    W = np.array([1, 0.5])
    dist = np.inf
    eps = 1e-8
    m = 10000
    
    for i in range(m):
        y_pred = np.dot(W, X)
        err = calc_mse(y, y_pred)
        W -= (alpha * (1/n * 2 * (X @ (y_pred - y))))
        if i % 100 == 0:
            print(i, W, err, alpha)
        if abs(dist - err) < eps:
            print(i, W, err, alpha)
            if i < best_i:
                best_i, best_a , best_err= i, alpha, err
            break
        if np.isinf(err):
            print("Ошибка переполнения")
            break
        dist = err
    alpha /= 2
print(best_i, best_a, best_err)

0 [12.02 34.5 ] 3173.15 0.1
100 [1.21092169e+31 6.55765787e+31] 1.64995985315541e+64 0.1
200 [3.40912000e+61 1.84618401e+62] 1.3077533388173702e+125 0.1
300 [9.59772979e+91 5.19758039e+92] 1.0365214595478379e+186 0.1
400 [2.70205851e+122 1.46328003e+123] 8.215438678021364e+246 0.1
500 [7.60713247e+152 4.11958697e+153] inf 0.1
Ошибка переполнения
0 [ 4.67333333 11.83333333] 3173.15 0.03333333333333333
100 [45.98140701  4.14167223] 46.83999216328279 0.03333333333333333
200 [47.19539555  3.91749996] 45.93827904612829 0.03333333333333333
300 [47.2310632   3.91091365] 45.93750067248549 0.03333333333333333
323 [47.23166319  3.91080286] 45.93750013273653 0.03333333333333333
0 [2.22444444 4.27777778] 3173.15 0.011111111111111112
100 [33.59535003  6.4288524 ] 148.26473814636546 0.011111111111111112
200 [42.96589566  4.69850947] 55.952680934707665 0.011111111111111112
300 [45.89745475  4.15717465] 46.91772629137573 0.011111111111111112
400 [46.81458798  3.98781899] 46.033438714294654 0.011111111

<ipython-input-5-bd4c0e5d829e>:2: RuntimeWarning: overflow encountered in square
  err = np.mean((y - y_pred)**2)


 [29.31081409  7.22002474] 219.05045212898736 0.0012345679012345679
800 [31.47110141  6.82111116] 179.83083859796838 0.0012345679012345679
900 [33.37098163  6.47028374] 149.4966265727611 0.0012345679012345679
1000 [35.04184494  6.16174605] 126.03478347065007 0.0012345679012345679
1100 [36.5112977   5.89040036] 107.88833940641447 0.0012345679012345679
1200 [37.8036185   5.65176343] 93.85306388508052 0.0012345679012345679
1300 [38.94015931  5.44189248] 82.99755091171572 0.0012345679012345679
1400 [39.93969831  5.25731996] 74.60140922316998 0.0012345679012345679
1500 [40.81875006  5.09499633] 68.10745583496069 0.0012345679012345679
1600 [41.59183845  4.95223964] 63.08474038153207 0.0012345679012345679
1700 [42.27173659  4.82669125] 59.19994647895853 0.0012345679012345679
1800 [42.86967789  4.71627681] 56.19527225335211 0.0012345679012345679
1900 [43.39554168  4.61917204] 53.87132214726224 0.0012345679012345679
2000 [43.85801636  4.53377255] 52.073874673732874 0.0012345679012345679
2100 [4

4700 [41.33497111  4.99967216] 64.65003984745151 0.00041152263374485596
4800 [41.5820363   4.95404968] 63.11494087091304 0.00041152263374485596
4900 [41.81875056  4.91033858] 61.70577503263273 0.00041152263374485596
5000 [42.04554755  4.86845878] 60.41221130148226 0.00041152263374485596
5100 [42.26284276  4.82833356] 59.22476616117865 0.00041152263374485596
5200 [42.47103427  4.78988941] 58.13473408368937 0.00041152263374485596
5300 [42.67050348  4.75305589] 57.1341237063106 0.00041152263374485596
5400 [42.86161582  4.71776554] 56.21559924451316 0.00041152263374485596
5500 [43.04472141  4.68395369] 55.37242671103362 0.00041152263374485596
5600 [43.22015569  4.65155841] 54.59842454692887 0.00041152263374485596
5700 [43.38824006  4.62052034] 53.88791830265346 0.00041152263374485596
5800 [43.54928244  4.59078263] 53.23569903691235 0.00041152263374485596
5900 [43.70357787  4.5622908 ] 52.63698512829965 0.00041152263374485596
6000 [43.851409    4.53499265] 52.08738721975133 0.00041152263374

6300 [29.28112956  7.22550621] 219.22820809564718 0.00013717421124828533
6400 [29.53536743  7.17855924] 214.35438044024204 0.00013717421124828533
6500 [29.78600455  7.13227718] 209.61762993268627 0.00013717421124828533
6600 [30.03309194  7.08665061] 205.01410125724254 0.00013717421124828533
6700 [30.27667986  7.04167023] 200.5400475295173 0.00013717421124828533
6800 [30.51681788  6.99732691] 196.19182724681428 0.00013717421124828533
6900 [30.75355485  6.95361162] 191.9659013242561 0.00013717421124828533
7000 [30.98693895  6.91051546] 187.8588302142694 0.00013717421124828533
7100 [31.21701766  6.86802967] 183.86727110708108 0.00013717421124828533
7200 [31.44383779  6.82614559] 179.98797520995547 0.00013717421124828533
7300 [31.6674455   6.78485472] 176.21778510294806 0.00013717421124828533
7400 [31.88788628  6.74414864] 172.55363216903294 0.00013717421124828533
7500 [32.10520499  6.70401908] 168.9925340965081 0.00013717421124828533
7600 [32.31944585  6.66445787] 165.53159245164267 0.000

8200 [17.37229501  9.42441835] 525.327715866487 4.5724737082761774e-05
8300 [17.51392935  9.39828321] 520.7908174715634 4.5724737082761774e-05
8400 [17.65489146  9.37226979] 516.2968558206682 4.5724737082761774e-05
8500 [17.79518461  9.34637779] 511.84542456083636 4.5724737082761774e-05
8600 [17.93481201  9.3206069 ] 507.4361211857866 4.5724737082761774e-05
8700 [18.07377685  9.29495677] 503.0685469992765 4.5724737082761774e-05
8800 [18.21208231  9.269427  ] 498.7423070788589 4.5724737082761774e-05
8900 [18.34973155  9.24401719] 494.4570102400309 4.5724737082761774e-05
9000 [18.48672772  9.21872693] 490.2122690007503 4.5724737082761774e-05
9100 [18.62307393  9.19355576] 486.00769954631585 4.5724737082761774e-05
9200 [18.75877328  9.16850325] 481.84292169460105 4.5724737082761774e-05
9300 [18.89382887  9.14356891] 477.717558861627 4.5724737082761774e-05
9400 [19.02824375  9.11875228] 473.63123802747606 4.5724737082761774e-05
9500 [19.16202099  9.09405286] 469.5835897025353 4.57247370827

100 [1.05628936 0.67310337] 3109.212780054802 5.0805263425290855e-06
200 [1.11151323 0.84181006] 3047.173246161759 5.0805263425290855e-06
300 [1.16623897 1.00788808] 2986.9741536444176 5.0805263425290855e-06
400 [1.22047399 1.17137743] 2928.5599856299286 5.0805263425290855e-06
500 [1.27422558 1.33231751] 2871.876900898076 5.0805263425290855e-06
600 [1.32750089 1.49074713] 2816.872683304415 5.0805263425290855e-06
700 [1.380307  1.6467045] 2763.4966927299793 5.0805263425290855e-06
800 [1.43265085 1.80022723] 2711.699817511495 5.0805263425290855e-06
900 [1.48453931 1.9513524 ] 2661.434428307423 5.0805263425290855e-06
1000 [1.53597911 2.1001165 ] 2612.6543333564673 5.0805263425290855e-06
1100 [1.58697689 2.24655546] 2565.3147350865474 5.0805263425290855e-06
1200 [1.63753921 2.39070468] 2519.3721880334456 5.0805263425290855e-06
1300 [1.68767249 2.53259901] 2474.7845580296253 5.0805263425290855e-06
1400 [1.7373831  2.67227277] 2431.5109826248636 5.0805263425290855e-06
1500 [1.78667727 2.8097

1900 [1.34479229 1.5419113 ] 2798.9192393373837 1.6935087808430284e-06
2000 [1.3623948  1.59389962] 2781.128562167039 1.6935087808430284e-06
2100 [1.3799457  1.64561604] 2763.5151140889116 1.6935087808430284e-06
2200 [1.39744524 1.69706196] 2746.0770950845445 1.6935087808430284e-06
2300 [1.41489368 1.74823874] 2728.812723429666 1.6935087808430284e-06
2400 [1.43229126 1.79914776] 2711.720235508254 1.6935087808430284e-06
2500 [1.44963824 1.84979039] 2694.7978856284894 1.6935087808430284e-06
2600 [1.46693488 1.90016799] 2678.0439458405867 1.6935087808430284e-06
2700 [1.48418141 1.9502819 ] 2661.456705756471 1.6935087808430284e-06
2800 [1.5013781  2.00013347] 2645.0344723712833 1.6935087808430284e-06
2900 [1.51852518 2.04972404] 2628.775569886708 1.6935087808430284e-06
3000 [1.53562291 2.09905494] 2612.6783395360985 1.6935087808430284e-06
3100 [1.55267152 2.1481275 ] 2596.741139411365 1.6935087808430284e-06
3200 [1.56967126 2.19694302] 2580.9623442916386 1.6935087808430284e-06
3300 [1.5866

3800 [1.23197228 1.20587621] 2915.832138557493 5.645029269476761e-07
3900 [1.23795676 1.22382217] 2909.492391498287 5.645029269476761e-07
4000 [1.24393531 1.24173682] 2903.173794435179 5.645029269476761e-07
4100 [1.24990793 1.25962021] 2896.876275519117 5.645029269476761e-07
4200 [1.25587462 1.27747239] 2890.5997631452774 5.645029269476761e-07
4300 [1.26183542 1.29529342] 2884.3441859522477 5.645029269476761e-07
4400 [1.26779031 1.31308336] 2878.109472821187 5.645029269476761e-07
4500 [1.27373931 1.33084225] 2871.895552875007 5.645029269476761e-07
4600 [1.27968243 1.34857014] 2865.7023554775524 5.645029269476761e-07
4700 [1.28561968 1.3662671 ] 2859.529810232774 5.645029269476761e-07
4800 [1.29155107 1.38393316] 2853.3778469839226 5.645029269476761e-07
4900 [1.29747661 1.4015684 ] 2847.2463958127273 5.645029269476761e-07
5000 [1.30339631 1.41917285] 2841.1353870385888 5.645029269476761e-07
5100 [1.30931018 1.43674657] 2835.0447512177716 5.645029269476761e-07
5200 [1.31521823 1.45428961

6000 [1.12318364 0.87730764] 3033.6483788574615 1.8816764231589204e-07
6100 [1.12521536 0.88348419] 3031.4015484712045 1.8816764231589204e-07
6200 [1.12724639 0.88965716] 3029.157222175203 1.8816764231589204e-07
6300 [1.12927675 0.89582653] 3026.9153971305795 1.8816764231589204e-07
6400 [1.13130642 0.90199232] 3024.676070501676 1.8816764231589204e-07
6500 [1.13333542 0.90815453] 3022.439239456052 1.8816764231589204e-07
6600 [1.13536374 0.91431315] 3020.2049011644785 1.8816764231589204e-07
6700 [1.13739137 0.92046818] 3017.9730528009395 1.8816764231589204e-07
6800 [1.13941833 0.92661964] 3015.743691542622 1.8816764231589204e-07
6900 [1.14144461 0.93276752] 3013.516814569914 1.8816764231589204e-07
7000 [1.14347022 0.93891183] 3011.2924190664053 1.8816764231589204e-07
7100 [1.14549514 0.94505256] 3009.070502218877 1.8816764231589204e-07
7200 [1.14751939 0.95118973] 3006.851061217306 1.8816764231589204e-07
7300 [1.14954297 0.95732332] 3004.6340932548505 1.8816764231589204e-07
7400 [1.15156

3500 [1.00806101 0.52485895] 3163.8224222056797 2.090751581287689e-08
3600 [1.0082911  0.52556818] 3163.5565145371975 2.090751581287689e-08
3700 [1.00852119 0.52627736] 3163.29063984818 2.090751581287689e-08
3800 [1.00875126 0.5269865 ] 3163.0247981344705 2.090751581287689e-08
3900 [1.00898133 0.52769559] 3162.758989391913 2.090751581287689e-08
4000 [1.00921139 0.52840463] 3162.49321361635 2.090751581287689e-08
4100 [1.00944144 0.52911363] 3162.22747080363 2.090751581287689e-08
4200 [1.00967148 0.52982258] 3161.9617609495954 2.090751581287689e-08
4300 [1.00990151 0.53053149] 3161.696084050094 2.090751581287689e-08
4400 [1.01013153 0.53124035] 3161.430440100972 2.090751581287689e-08
4500 [1.01036155 0.53194917] 3161.1648290980775 2.090751581287689e-08
4600 [1.01059155 0.53265793] 3160.8992510372564 2.090751581287689e-08
4700 [1.01082155 0.53336666] 3160.6337059143575 2.090751581287689e-08
4800 [1.01105154 0.53407534] 3160.3681937252295 2.090751581287689e-08
4900 [1.01128152 0.53478397] 

600 [1.00015385 0.50047468] 3172.971953320001 2.323057312541877e-09
700 [1.00017945 0.50055366] 3172.942280304766 2.323057312541877e-09
800 [1.00020505 0.50063264] 3172.9126076984967 2.323057312541877e-09
900 [1.00023065 0.50071162] 3172.882935501187 2.323057312541877e-09
1000 [1.00025625 0.5007906 ] 3172.853263712831 2.323057312541877e-09
1100 [1.00028185 0.50086958] 3172.8235923334223 2.323057312541877e-09
1200 [1.00030745 0.50094856] 3172.793921362956 2.323057312541877e-09
1300 [1.00033305 0.50102753] 3172.7642508014264 2.323057312541877e-09
1400 [1.00035865 0.50110651] 3172.7345806488283 2.323057312541877e-09
1500 [1.00038425 0.50118549] 3172.704910905155 2.323057312541877e-09
1600 [1.00040984 0.50126446] 3172.6752415704013 2.323057312541877e-09
1700 [1.00043544 0.50134344] 3172.645572644561 2.323057312541877e-09
1800 [1.00046104 0.50142241] 3172.6159041276287 2.323057312541877e-09
1900 [1.00048664 0.50150138] 3172.5862360195983 2.323057312541877e-09
2000 [1.00051224 0.50158036] 31

8200 [1.00069978 0.50215895] 3172.3390177056262 7.743524375139591e-10
8300 [1.00070831 0.50218527] 3172.3291295631307 7.743524375139591e-10
8400 [1.00071685 0.50221159] 3172.319241466063 7.743524375139591e-10
8500 [1.00072538 0.50223791] 3172.3093534144223 7.743524375139591e-10
8600 [1.00073391 0.50226424] 3172.299465408208 7.743524375139591e-10
8700 [1.00074244 0.50229056] 3172.289577447421 7.743524375139591e-10
8800 [1.00075098 0.50231688] 3172.2796895320607 7.743524375139591e-10
8900 [1.00075951 0.5023432 ] 3172.2698016621266 7.743524375139591e-10
9000 [1.00076804 0.50236953] 3172.2599138376186 7.743524375139591e-10
9100 [1.00077657 0.50239585] 3172.250026058536 7.743524375139591e-10
9200 [1.0007851  0.50242217] 3172.2401383248807 7.743524375139591e-10
9300 [1.00079364 0.50244849] 3172.2302506366505 7.743524375139591e-10
9400 [1.00080217 0.50247482] 3172.2203629938444 7.743524375139591e-10
9500 [1.0008107  0.50250114] 3172.210475396464 7.743524375139591e-10
9600 [1.00081923 0.502527

200 [1.00000191 0.50000588] 3173.1478018031476 8.603915972377323e-11
300 [1.00000285 0.50000881] 3173.1467027055623 8.603915972377323e-11
400 [1.0000038  0.50001173] 3173.1456036085383 8.603915972377323e-11
500 [1.00000475 0.50001466] 3173.144504512076 8.603915972377323e-11
600 [1.0000057  0.50001758] 3173.143405416173 8.603915972377323e-11
700 [1.00000665 0.50002051] 3173.142306320833 8.603915972377323e-11
800 [1.00000759 0.50002343] 3173.1412072260537 8.603915972377323e-11
900 [1.00000854 0.50002636] 3173.140108131835 8.603915972377323e-11
1000 [1.00000949 0.50002928] 3173.139009038177 8.603915972377323e-11
1100 [1.00001044 0.50003221] 3173.137909945081 8.603915972377323e-11
1200 [1.00001139 0.50003513] 3173.1368108525453 8.603915972377323e-11
1300 [1.00001234 0.50003806] 3173.1357117605708 8.603915972377323e-11
1400 [1.00001328 0.50004098] 3173.134612669158 8.603915972377323e-11
1500 [1.00001423 0.50004391] 3173.1335135783047 8.603915972377323e-11
1600 [1.00001518 0.50004683] 3173.1

Методом пошагового подбора получаем $i = 323$, $\alpha = 0.33333$, $err = 45.9375001$